triton-vm/src/table/hash_table.rs

```diff
@@ -765,36 +781,43 @@ impl ExtHashTable {
-            .sum::<ConstraintCircuitMonad<_>>();
-
-        let capacity_doesnt_change_at_section_start_when_program_hashing_or_absorbing =
+        let randomized_sum_of_capacity_differences = izip!(
+            &state_weights[RATE..],
+            &state_current[RATE..],
+            &state_next[RATE..]
+        )
+        .map(|(weight, current, next)| weight.clone() * (next.clone() - current.clone()))
+        .sum::<ConstraintCircuitMonad<_>>();
+
+        // If the next round number is 0, then
+        // - the next mode is either `Hash` or `Pad`, or
+        // - the next instruction is `sponge_init`, or
+        // - the capacity does not change, or
+        // - the current mode is `ProgramHashing` and (!) the next mode is `sponge`.
+        //
+        // The “and” marked with an exclamation mark requires above constraint to be
+        // expressed through two polynomials.
+        let capacity_doesnt_change_at_section_start_when_program_hashing_or_ =
             Self::round_number_deselector(circuit_builder, &round_number_next, 0)
                 * Self::select_mode(circuit_builder, &mode_next, HashTableMode::Hash)
                 * Self::select_mode(circuit_builder, &mode_next, HashTableMode::Pad)
                 * (ci_next.clone() - opcode_sponge_init.clone())
                 * randomized_sum_of_capacity_differences.clone();
 
-        let difference_of_state_registers = state_current
-            .iter()
-            .zip_eq(state_next.iter())
-            .map(|(current, next)| next.clone() - current.clone())
-            .collect_vec();
-        let randomized_sum_of_state_differences = state_weights
-            .iter()
-            .zip_eq(difference_of_state_registers.iter())
-            .map(|(weight, state_difference)| weight.clone() * state_difference.clone())
-            .sum();
-        let if_round_number_next_is_0_and_ci_next_is_squeeze_then_state_doesnt_change =
+        let capacity_doesnt_change_at_section_start_when_program_hashing_or_absorbing_0 =
+            capacity_doesnt_change_at_section_start_when_program_hashing_or_.clone()
+                * Self::select_mode(circuit_builder, &mode, HashTableMode::ProgramHashing);
+        let capacity_doesnt_change_at_section_start_when_program_hashing_or_absorbing_1 =
+            capacity_doesnt_change_at_section_start_when_program_hashing_or_
+                * Self::select_mode(circuit_builder, &mode_next, HashTableMode::Sponge);
+
+        let randomized_sum_of_state_differences =
+            izip!(&state_weights, &state_current, &state_next)
+                .map(|(weight, curr, next)| weight.clone() * (next.clone() - curr.clone()))
+                .sum();
+        let if_next_round_is_0_and_ci_next_is_squeeze_in_sponge_section_then_state_doesnt_change =
             Self::round_number_deselector(circuit_builder, &round_number_next, 0)
+                * Self::select_mode(circuit_builder, &mode, HashTableMode::ProgramHashing)
```

Comment:
I would expect there to be a *deselector* for mode `Sponge`. Can't you transition from something other than `ProgramHashing` to `Sponge`?

Reply:
> The four separate modes are `program_hashing`, `sponge`, `hash`, and `pad`, and they evolve in that order.

Got it!